File: rdna3_kawpow/miner.py
"""KawPow miner orchestration on top of the Vulkan host.

Responsibilities:
  * per-epoch: build the light cache (host) and generate the full DAG on the GPU
    (chunked, to stay under the OS GPU watchdog),
  * per-period: regenerate + compile the ProgPoW search kernel,
  * per-batch: dispatch the search over a nonce range and collect solutions.

All hashing is on the GPU; this code only feeds it.
"""

import struct
import time

from . import dagcache, ethash, keccak, shader_compiler as sc
from .constants import PROGPOW_PERIOD, KAWPOW_EPOCH_LENGTH
from .vkhost import VulkanDevice, ComputePipeline

MAX_OUTPUTS = 16
SEARCH_LOCAL_SIZE = 128          # multiple of 16 lanes and of wave32
DAG_LOCAL_SIZE = 64
# Host-visible staging window for streaming the device-local DAG to/from disk.
DAG_STAGING_BYTES = 64 << 20     # 64 MiB
# IMPORTANT: keep every single GPU dispatch well under the OS GPU watchdog (TDR,
# ~2 s on Windows). A too-large dispatch is killed and can hang the machine. These
# are deliberately small; the miner splits big nonce ranges / the DAG into many
# such chunks. The search auto-calibrates a safe batch at startup.
DAG_CHUNK_ITEMS = 1 << 15        # ~32k items per DAG dispatch
SEARCH_CHUNK_NONCES = 1 << 20    # default safe nonce batch (auto-tuned by calibrate())
WATCHDOG_TARGET_S = 0.20         # aim for ~200 ms dispatches, far below the TDR limit


class Solution:
    def __init__(self, nonce, mix_words):
        self.nonce = nonce
        self.mix_words = mix_words            # 8 "my-domain" digest words

    def mix_hash(self):
        """Canonical 32-byte mix hash (big-endian word read)."""
        return struct.pack(">8I", *[keccak._swab32(w) for w in self.mix_words])


class VulkanMiner:
    def __init__(self, device_index=None, variant=keccak.KAWPOW,
                 epoch_length=KAWPOW_EPOCH_LENGTH, dag_cache=True, cache_dir=None):
        self.dev = VulkanDevice(device_index)
        self.variant = variant
        self.epoch_length = epoch_length
        # Disk cache of the light cache + full DAG, so restarts within an epoch
        # skip both the host mkcache and the GPU DAG build.
        self._cache = dagcache.Cache(cache_dir) if dag_cache else None
        # Standard KawPow uses 512 DAG parents; the 0.9.2 reference uses 256.
        self.parents = 512 if variant == keccak.KAWPOW else 256
        self._epoch = None
        self._light = None
        self._dag_buf = None
        self._light_buf = None
        self._dag_elements = None
        self._period = None
        self._search_pipe = None
        self._safe_batch = SEARCH_CHUNK_NONCES
        # Persistent small buffers.
        self._hdr_buf = self.dev.make_buffer(32)
        self._out_size = 16 + MAX_OUTPUTS * 9 * 4
        self._out_buf = self.dev.make_buffer(self._out_size)

    # --- per-epoch DAG ---
    def ensure_epoch(self, block, seed=None, log=print):
        epoch = block // self.epoch_length
        if epoch == self._epoch and self._dag_buf is not None and self._light is not None:
            return
        eff_seed = seed if seed is not None else ethash.seed_hash(epoch)
        cache = self._cache

        # 1) Light cache: reload from disk if present (and still needed for the
        #    host solution re-check), otherwise build it and cache it.
        light_bytes = cache.load_light(self.variant, epoch, self.parents,
                                       eff_seed) if cache else None
        self._light = None
        if light_bytes is not None:
            t0 = time.time()
            try:
                self._light = ethash.LightCache(
                    block, self.epoch_length, seed=eff_seed,
                    parents=self.parents, precomputed_cache=light_bytes)
                log(f"Epoch {epoch}: loaded {len(light_bytes)//(1<<20)} MiB light "
                    f"cache from disk in {time.time()-t0:.1f}s")
            except ValueError as e:
                log(f"Epoch {epoch}: cached light cache rejected ({e}); rebuilding")
                light_bytes = None
        if self._light is None:
            t0 = time.time()
            log(f"Epoch {epoch}: building light cache (host)...")
            self._light = ethash.LightCache(block, self.epoch_length, seed=eff_seed,
                                            parents=self.parents)
            light_bytes = b"".join(self._light.cache)
            log(f"  light={len(light_bytes)//(1<<20)} MiB cache in {time.time()-t0:.1f}s")
            if cache:
                # Caching is a pure optimization: never let a write failure
                # (full disk, etc.) abort an otherwise-good epoch setup.
                try:
                    cache.save_light(self.variant, epoch, self.parents,
                                     eff_seed, light_bytes)
                except Exception as e:
                    log(f"  (light cache save skipped: {e})")

        full_size = self._light.full_size

        # (Re)allocate the device-local DAG buffer. transfer=True lets us stage it
        # to/from disk for caching.
        if self._dag_buf:
            self._dag_buf.destroy()
        self._dag_buf = self.dev.make_buffer(full_size, host_visible=False,
                                             transfer=True)

        # 2) DAG: reload from disk if present, else generate on the GPU and cache it.
        loaded = False
        if cache and cache.has_dag(self.variant, epoch, self.parents,
                                   eff_seed, full_size):
            try:
                self._load_dag_from_disk(cache, epoch, log)
                loaded = True
            except Exception as e:
                # Fall back to GPU generation, which fully overwrites the buffer.
                log(f"  DAG cache load failed ({e}); regenerating")
        if not loaded:
            self._generate_dag(light_bytes, log)
            if cache:
                # Best-effort: the DAG is already on the GPU and usable; a save
                # failure must not abort epoch setup (or it would loop forever).
                try:
                    self._save_dag_to_disk(cache, epoch, eff_seed, log)
                except Exception as e:
                    log(f"  (DAG cache save skipped: {e})")

        self._dag_elements = self._light.dag_elements
        self._epoch = epoch
        self._period = None  # force kernel recompile (dag_elements may change)

    def _generate_dag(self, light_bytes, log):
        """Generate the full DAG on the GPU from the light cache (watchdog-safe)."""
        log(f"  generating DAG={self._light.full_size//(1<<20)} MiB on GPU...")
        if self._light_buf:
            self._light_buf.destroy()
        self._light_buf = self.dev.make_buffer(len(light_bytes))
        self._light_buf.write(light_bytes)

        spv = sc.compile_dag(len(self._light.cache), self._light.dag_items, self.parents)
        dag_pipe = ComputePipeline(self.dev, spv, num_bindings=2, push_const_size=4,
                                   local_size=DAG_LOCAL_SIZE, required_subgroup_size=32)
        dag_pipe.bind([self._light_buf, self._dag_buf])

        # Adaptive, watchdog-safe DAG generation: start with tiny dispatches and
        # grow the chunk only while each stays fast, so no dispatch ever runs long
        # enough to trip the OS GPU watchdog (which can hang the machine).
        t1 = time.time()
        items = self._light.dag_items
        start = 0
        chunk = 1 << 12                       # 4096 items -- definitely safe
        last_log = t1
        while start < items:
            n = min(chunk, items - start)
            groups = (n + DAG_LOCAL_SIZE - 1) // DAG_LOCAL_SIZE
            t0 = time.time()
            self.dev.dispatch(dag_pipe, groups, struct.pack("<I", start))
            dt = time.time() - t0
            start += n
            if dt < 0.12 and chunk < (1 << 19):
                chunk <<= 1
            elif dt > 0.30 and chunk > (1 << 12):
                chunk >>= 1
            if time.time() - last_log > 3.0:
                log(f"  DAG {100.0*start/items:5.1f}%  ({n} items, {dt*1000:.0f} ms/dispatch)")
                last_log = time.time()
        log(f"  DAG generated in {time.time()-t1:.1f}s")
        # The light cache on the GPU is only needed during generation.
        self._light_buf.destroy()
        self._light_buf = None

    def _load_dag_from_disk(self, cache, epoch, log):
        """Upload a cached DAG into the device-local buffer via a staging window."""
        full = self._light.full_size
        t0 = time.time()
        reader = stage = None
        try:
            reader = cache.open_dag_read(self.variant, epoch, self.parents)
            stage = self.dev.make_buffer(min(DAG_STAGING_BYTES, full),
                                         host_visible=True, storage=False, transfer=True)
            pos = 0
            while pos < full:
                n = min(DAG_STAGING_BYTES, full - pos)
                data = reader.read(n)
                if len(data) != n:
                    raise RuntimeError("short read from DAG cache")
                stage.write(data)
                self.dev.copy_buffer(stage, self._dag_buf, n, 0, pos)
                pos += n
        finally:
            if reader is not None:
                reader.close()
            if stage is not None:
                stage.destroy()
        log(f"  loaded DAG={full//(1<<20)} MiB from disk in {time.time()-t0:.1f}s")

    def _save_dag_to_disk(self, cache, epoch, seed, log):
        """Stream the device-local DAG out to disk via a staging window."""
        full = self._light.full_size
        t0 = time.time()
        w = stage = None
        try:
            w = cache.open_dag_write(self.variant, epoch, self.parents, seed, full)
            stage = self.dev.make_buffer(min(DAG_STAGING_BYTES, full),
                                         host_visible=True, storage=False, transfer=True)
            pos = 0
            while pos < full:
                n = min(DAG_STAGING_BYTES, full - pos)
                self.dev.copy_buffer(self._dag_buf, stage, n, pos, 0)
                w.write(stage.read(n))
                pos += n
            w.commit()
        except BaseException:
            if w is not None:
                w.abort()
            raise
        finally:
            if stage is not None:
                stage.destroy()
        cache.prune_dag(self.variant)
        log(f"  saved DAG={full//(1<<20)} MiB to disk in {time.time()-t0:.1f}s")

    def setup_benchmark_dag(self, block, log=print):
        """Allocate an UNINITIALIZED full-size DAG buffer for throughput testing.

        Hashes are garbage but timing is representative (same memory footprint and
        access pattern). Avoids the multi-second DAG build entirely.
        """
        epoch, full_size, dag_items, dag_elements = ethash.dag_sizing(
            block, self.epoch_length)
        if self._dag_buf is None or self._epoch != epoch:
            if self._dag_buf:
                self._dag_buf.destroy()
            log(f"Allocating {full_size//(1<<20)} MiB device-local DAG buffer "
                f"(uninitialized, benchmark only)...")
            self._dag_buf = self.dev.make_buffer(full_size, host_visible=False)
        self._dag_elements = dag_elements
        self._epoch = epoch
        self._period = None

    # --- per-period kernel ---
    def ensure_period(self, block, log=print):
        period = block // PROGPOW_PERIOD
        if period == self._period and self._search_pipe is not None:
            return
        from .constants import PROGPOW_CNT_CACHE as cc, PROGPOW_CNT_MATH as cm
        t0 = time.time()
        spv = sc.compile_search(period, self._dag_elements, self.variant,
                                MAX_OUTPUTS, cc, cm)
        self._search_pipe = ComputePipeline(
            self.dev, spv, num_bindings=3, push_const_size=16,
            local_size=SEARCH_LOCAL_SIZE, required_subgroup_size=32)
        self._search_pipe.bind([self._hdr_buf, self._dag_buf, self._out_buf])
        self._period = period
        log(f"Compiled period {period} kernel in {(time.time()-t0)*1000:.0f} ms")

    # --- search ---
    def search(self, header_bytes, target, start_nonce, num_nonces):
        """Dispatch the search over [start_nonce, start_nonce+num_nonces).

        target is the 64-bit boundary (upper bits). Returns (solutions, hashes).
        """
        self._hdr_buf.write(header_bytes)
        self._out_buf.write(b"\x00" * 16)  # zero count/hashCount/abort
        groups = (num_nonces + SEARCH_LOCAL_SIZE - 1) // SEARCH_LOCAL_SIZE
        push = struct.pack("<4I", start_nonce & 0xFFFFFFFF, start_nonce >> 32,
                           target & 0xFFFFFFFF, (target >> 32) & 0xFFFFFFFF)
        self.dev.dispatch(self._search_pipe, groups, push)

        raw = self._out_buf.read(self._out_size)
        count, hash_count, _abort, _pad = struct.unpack("<4I", raw[:16])
        sols = []
        for s in range(min(count, MAX_OUTPUTS)):
            base = 16 + s * 9 * 4
            v = struct.unpack("<9I", raw[base:base + 36])
            sols.append(Solution(start_nonce + v[0], v[1:9]))
        return sols, hash_count

    def calibrate(self, header, log=print):
        """Find a per-dispatch nonce batch that runs ~WATCHDOG_TARGET_S.

        Ramps up from a tiny batch, doubling and measuring, and STOPS as soon as a
        dispatch reaches the target -- so no single dispatch ever runs long enough
        to trip the GPU watchdog.
        """
        batch = 1 << 13
        per_nonce = None
        while batch <= (1 << 24):
            t0 = time.time()
            _, hashes = self.search(header, 0, 0, batch)
            dt = time.time() - t0
            if hashes:
                per_nonce = dt / hashes
            log(f"  calibrate batch={batch} -> {dt*1000:.1f} ms "
                f"({hashes/dt/1e6:.1f} MH/s)")
            if dt >= WATCHDOG_TARGET_S:
                break
            batch <<= 1
        if per_nonce:
            safe = int(WATCHDOG_TARGET_S / per_nonce)
            # round down to a multiple of the local size
            self._safe_batch = max(SEARCH_LOCAL_SIZE,
                                   (safe // SEARCH_LOCAL_SIZE) * SEARCH_LOCAL_SIZE)
        log(f"  safe batch = {self._safe_batch} nonces (~{WATCHDOG_TARGET_S*1000:.0f} ms/dispatch)")
        return self._safe_batch

    def benchmark(self, block, seconds=10.0, use_real_dag=False, log=print):
        """Measure sustained hashrate (MH/s) using watchdog-safe dispatches."""
        if use_real_dag:
            self.ensure_epoch(block, log=log)
        else:
            self.setup_benchmark_dag(block, log)
        self.ensure_period(block, log)

        header = bytes(range(32))
        log("Calibrating safe dispatch size...")
        self.calibrate(header, log)

        start = 0
        total = 0
        t0 = time.time()
        while time.time() - t0 < seconds:
            _, hashes = self.search(header, 0, start, self._safe_batch)
            total += hashes
            start += self._safe_batch
        dt = time.time() - t0
        mhs = total / dt / 1e6
        log(f"Benchmark: {total/1e6:.1f} Mhashes in {dt:.1f}s = {mhs:.2f} MH/s")
        return mhs

File: rust/src/miner.rs
//! KawPow miner orchestration on top of the Vulkan host (port of
//! `rdna3_kawpow/miner.py`). Per-epoch: build the light cache (host) + generate
//! the full DAG on the GPU (chunked, watchdog-safe). Per-period: regenerate +
//! compile the search kernel. Per-batch: dispatch the search over a nonce range.
//!
//! Phase 2 scope: epoch DAG, benchmark DAG, period kernel, search, calibration,
//! benchmark. The on-disk DAG cache and the live mining loop land in Phase 3.

use std::path::PathBuf;
use std::time::Instant;

use anyhow::{anyhow, Result};

use crate::constants::PROGPOW_PERIOD;
use crate::dagcache::Cache;
use crate::ethash::{self, LightCache};
use crate::keccak::{self, Variant};
use crate::shader_compiler as sc;
use crate::vkhost::{Buffer, ComputePipeline, VulkanDevice};

pub const MAX_OUTPUTS: u32 = 16;
pub const SEARCH_LOCAL_SIZE: u32 = 256; // wave32 multiple, tuned on gfx1100
const DAG_LOCAL_SIZE: u32 = 64;
// Keep every single GPU dispatch well under the OS GPU watchdog (TDR, ~2 s on
// Windows). The search auto-calibrates a safe batch at startup.
const SEARCH_CHUNK_NONCES: u32 = 1 << 20;
const WATCHDOG_TARGET_S: f64 = 0.20; // aim for ~200 ms dispatches
// Host-visible staging window for streaming the device-local DAG to/from disk.
const DAG_STAGING_BYTES: u64 = 64 << 20; // 64 MiB

fn variant_str(v: Variant) -> &'static str {
    match v {
        Variant::Kawpow => "kawpow",
        Variant::Vanilla => "vanilla",
    }
}

/// A found solution: nonce + the 8 "my-domain" digest words.
pub struct Solution {
    pub nonce: u64,
    pub mix_words: [u32; 8],
}

impl Solution {
    /// Canonical 32-byte mix hash (big-endian word read).
    pub fn mix_hash(&self) -> [u8; 32] {
        let mut out = [0u8; 32];
        for (i, &w) in self.mix_words.iter().enumerate() {
            out[i * 4..i * 4 + 4].copy_from_slice(&keccak::swab32(w).to_be_bytes());
        }
        out
    }
}

pub struct VulkanMiner {
    pub dev: VulkanDevice,
    pub variant: Variant,
    pub epoch_length: u64,
    pub local_size: u32,
    parents: u32,

    cache: Option<Cache>,
    epoch: Option<u64>,
    light: Option<LightCache>,
    dag_buf: Option<Buffer>,
    dag_elements: Option<u64>,
    period: Option<u64>,
    search_pipe: Option<ComputePipeline>,
    safe_batch: u32,

    hdr_buf: Buffer,
    out_buf: Buffer,
    out_size: usize,
}

impl VulkanMiner {
    pub fn new(
        device_index: Option<usize>,
        variant: Variant,
        epoch_length: u64,
        local_size: u32,
        dag_cache: bool,
        cache_dir: Option<PathBuf>,
    ) -> Result<Self> {
        let dev = VulkanDevice::new(device_index)?;
        let parents = if variant == Variant::Kawpow { 512 } else { 256 };
        let out_size = 16 + (MAX_OUTPUTS as usize) * 9 * 4;
        let hdr_buf = dev.make_buffer(32, true, true, false)?;
        let out_buf = dev.make_buffer(out_size as u64, true, true, false)?;
        // Best-effort: a cache-dir failure degrades to "no cache", never aborts.
        let cache = if dag_cache {
            match Cache::new(cache_dir) {
                Ok(c) => Some(c),
                Err(e) => {
                    eprintln!("DAG cache disabled (cache dir error: {e})");
                    None
                }
            }
        } else {
            None
        };
        Ok(VulkanMiner {
            dev,
            variant,
            epoch_length,
            local_size,
            parents,
            cache,
            epoch: None,
            light: None,
            dag_buf: None,
            dag_elements: None,
            period: None,
            search_pipe: None,
            safe_batch: SEARCH_CHUNK_NONCES,
            hdr_buf,
            out_buf,
            out_size,
        })
    }

    pub fn light(&self) -> Option<&LightCache> {
        self.light.as_ref()
    }
    pub fn safe_batch(&self) -> u32 {
        self.safe_batch
    }

    // --- per-epoch DAG ---
    pub fn ensure_epoch(&mut self, block: u64, seed: Option<[u8; 32]>) -> Result<()> {
        let epoch = block / self.epoch_length;
        if self.epoch == Some(epoch) && self.dag_buf.is_some() && self.light.is_some() {
            return Ok(());
        }
        let eff_seed = seed.unwrap_or_else(|| ethash::seed_hash(epoch));
        let vstr = variant_str(self.variant);
        let e32 = epoch as u32;

        // 1) Light cache: reload from disk if present, else build it and cache it.
        let mut light: Option<LightCache> = None;
        if let Some(cache) = &self.cache {
            if let Some(bytes) = cache.load_light(vstr, e32, self.parents, &eff_seed) {
                let mib = bytes.len() / (1 << 20);
                match LightCache::from_precomputed(
                    block,
                    self.epoch_length,
                    eff_seed,
                    self.parents,
                    &bytes,
                ) {
                    Ok(lc) => {
                        println!("Epoch {epoch}: loaded {mib} MiB light cache from disk");
                        light = Some(lc);
                    }
                    Err(e) => println!("Epoch {epoch}: cached light cache rejected ({e}); rebuilding"),
                }
            }
        }
        let light = match light {
            Some(lc) => lc,
            None => {
                let t0 = Instant::now();
                println!("Epoch {epoch}: building light cache (host)...");
                let lc = LightCache::new(block, self.epoch_length, Some(eff_seed), self.parents);
                println!(
                    "  light={} MiB cache in {:.1}s",
                    lc.cache_size / (1 << 20),
                    t0.elapsed().as_secs_f64()
                );
                if let Some(cache) = &self.cache {
                    if let Err(e) = cache.save_light(vstr, e32, self.parents, &eff_seed, &lc.flatten_cache()) {
                        println!("  (light cache save skipped: {e})");
                    }
                }
                lc
            }
        };
        let full_size = light.full_size;

        // (Re)allocate the device-local DAG buffer (transfer=true for caching).
        self.dag_buf = None; // free the old one first
        self.dag_buf = Some(self.dev.make_buffer(full_size, false, true, true)?);

        // 2) DAG: reload from disk if present, else generate on the GPU and cache it.
        let mut loaded = false;
        if let Some(cache) = &self.cache {
            if cache.has_dag(vstr, e32, self.parents, &eff_seed, full_size) {
                match self.load_dag_from_disk(cache, e32, full_size) {
                    Ok(()) => loaded = true,
                    Err(e) => println!("  DAG cache load failed ({e}); regenerating"),
                }
            }
        }
        if !loaded {
            self.generate_dag(&light)?;
            if let Some(cache) = &self.cache {
                if let Err(e) = self.save_dag_to_disk(cache, e32, &eff_seed, full_size) {
                    println!("  (DAG cache save skipped: {e})");
                }
            }
        }

        self.dag_elements = Some(light.dag_elements);
        self.light = Some(light);
        self.epoch = Some(epoch);
        self.period = None; // force kernel recompile (dag_elements may change)
        Ok(())
    }

    /// Upload a cached DAG into the device-local buffer via a staging window.
    fn load_dag_from_disk(&self, cache: &Cache, epoch: u32, full: u64) -> Result<()> {
        use std::io::Read;
        let t0 = Instant::now();
        let mut reader = cache.open_dag_read(variant_str(self.variant), epoch, self.parents)?;
        let stage = self
            .dev
            .make_buffer(DAG_STAGING_BYTES.min(full), true, false, true)?;
        let dag = self.dag_buf.as_ref().unwrap();
        let mut buf = vec![0u8; DAG_STAGING_BYTES as usize];
        let mut pos = 0u64;
        while pos < full {
            let n = DAG_STAGING_BYTES.min(full - pos) as usize;
            reader.read_exact(&mut buf[..n])?;
            stage.write(&buf[..n], 0);
            self.dev.copy_buffer(&stage, dag, n as u64, 0, pos)?;
            pos += n as u64;
        }
        println!(
            "  loaded DAG={} MiB from disk in {:.1}s",
            full / (1 << 20),
            t0.elapsed().as_secs_f64()
        );
        Ok(())
    }

    /// Stream the device-local DAG out to disk via a staging window.
    fn save_dag_to_disk(&self, cache: &Cache, epoch: u32, seed: &[u8; 32], full: u64) -> Result<()> {
        let t0 = Instant::now();
        let mut w = cache.open_dag_write(variant_str(self.variant), epoch, self.parents, seed, full)?;
        let stage = self
            .dev
            .make_buffer(DAG_STAGING_BYTES.min(full), true, false, true)?;
        let dag = self.dag_buf.as_ref().unwrap();
        let res = (|| -> Result<()> {
            let mut pos = 0u64;
            while pos < full {
                let n = DAG_STAGING_BYTES.min(full - pos);
                self.dev.copy_buffer(dag, &stage, n, pos, 0)?;
                let data = stage.read(n as usize, 0);
                w.write_all(&data)?;
                pos += n;
            }
            Ok(())
        })();
        match res {
            Ok(()) => w.commit()?,
            Err(e) => {
                w.abort();
                return Err(e);
            }
        }
        cache.prune_dag(variant_str(self.variant));
        println!(
            "  saved DAG={} MiB to disk in {:.1}s",
            full / (1 << 20),
            t0.elapsed().as_secs_f64()
        );
        Ok(())
    }

    fn generate_dag(&self, light: &LightCache) -> Result<()> {
        println!("  generating DAG={} MiB on GPU...", light.full_size / (1 << 20));
        // Flatten the light cache to a byte buffer for upload.
        let mut light_bytes = Vec::with_capacity(light.cache.len() * 64);
        for item in &light.cache {
            light_bytes.extend_from_slice(item);
        }
        let light_buf = self.dev.make_buffer(light_bytes.len() as u64, true, true, false)?;
        light_buf.write(&light_bytes, 0);

        let spv = sc::compile_dag(
            light.cache.len() as u32,
            light.dag_items as u32,
            self.parents,
        )?;
        let dag_pipe = ComputePipeline::new(&self.dev, &spv, 2, 4, DAG_LOCAL_SIZE, 32)?;
        dag_pipe.bind(&[&light_buf, self.dag_buf.as_ref().unwrap()]);

        // Adaptive, watchdog-safe DAG generation: start tiny, grow only while each
        // dispatch stays fast, so no dispatch trips the OS GPU watchdog.
        let t1 = Instant::now();
        let items = light.dag_items;
        let mut start = 0u64;
        let mut chunk = 1u64 << 12; // 4096 items -- definitely safe
        let mut last_log = Instant::now();
        while start < items {
            let n = chunk.min(items - start);
            let groups = ((n + DAG_LOCAL_SIZE as u64 - 1) / DAG_LOCAL_SIZE as u64) as u32;
            let t0 = Instant::now();
            self.dev
                .dispatch(&dag_pipe, groups, &(start as u32).to_le_bytes())?;
            let dt = t0.elapsed().as_secs_f64();
            start += n;
            if dt < 0.12 && chunk < (1 << 19) {
                chunk <<= 1;
            } else if dt > 0.30 && chunk > (1 << 12) {
                chunk >>= 1;
            }
            if last_log.elapsed().as_secs_f64() > 3.0 {
                println!(
                    "  DAG {:5.1}%  ({} items, {:.0} ms/dispatch)",
                    100.0 * start as f64 / items as f64,
                    n,
                    dt * 1000.0
                );
                last_log = Instant::now();
            }
        }
        println!("  DAG generated in {:.1}s", t1.elapsed().as_secs_f64());
        Ok(())
    }

    /// Allocate an UNINITIALIZED full-size DAG buffer for throughput testing.
    pub fn setup_benchmark_dag(&mut self, block: u64) -> Result<()> {
        let (epoch, full_size, _dag_items, dag_elements) =
            ethash::dag_sizing(block, self.epoch_length);
        if self.dag_buf.is_none() || self.epoch != Some(epoch) {
            self.dag_buf = None;
            println!(
                "Allocating {} MiB device-local DAG buffer (uninitialized, benchmark only)...",
                full_size / (1 << 20)
            );
            self.dag_buf = Some(self.dev.make_buffer(full_size, false, true, false)?);
        }
        self.dag_elements = Some(dag_elements);
        self.epoch = Some(epoch);
        self.period = None;
        Ok(())
    }

    // --- per-period kernel ---
    pub fn ensure_period(&mut self, block: u64) -> Result<()> {
        let period = block / PROGPOW_PERIOD;
        if self.period == Some(period) && self.search_pipe.is_some() {
            return Ok(());
        }
        let de = self
            .dag_elements
            .ok_or_else(|| anyhow!("dag_elements unset (call ensure_epoch first)"))?;
        let t0 = Instant::now();
        let spv = sc::compile_search(period, de, self.variant, MAX_OUTPUTS)?;
        let pipe = ComputePipeline::new(&self.dev, &spv, 3, 16, self.local_size, 32)?;
        pipe.bind(&[
            &self.hdr_buf,
            self.dag_buf.as_ref().unwrap(),
            &self.out_buf,
        ]);
        self.search_pipe = Some(pipe);
        self.period = Some(period);
        println!(
            "Compiled period {period} kernel in {:.0} ms",
            t0.elapsed().as_secs_f64() * 1000.0
        );
        Ok(())
    }

    // --- search ---
    pub fn search(
        &self,
        header_bytes: &[u8],
        target: u64,
        start_nonce: u64,
        num_nonces: u64,
    ) -> Result<(Vec<Solution>, u64)> {
        let pipe = self
            .search_pipe
            .as_ref()
            .ok_or_else(|| anyhow!("no search pipeline (call ensure_period first)"))?;
        self.hdr_buf.write(header_bytes, 0);
        self.out_buf.write(&[0u8; 16], 0); // zero count/hashCount/abort/pad

        let groups = ((num_nonces + self.local_size as u64 - 1) / self.local_size as u64) as u32;
        let mut push = [0u8; 16];
        push[0..4].copy_from_slice(&(start_nonce as u32).to_le_bytes());
        push[4..8].copy_from_slice(&((start_nonce >> 32) as u32).to_le_bytes());
        push[8..12].copy_from_slice(&(target as u32).to_le_bytes());
        push[12..16].copy_from_slice(&((target >> 32) as u32).to_le_bytes());
        self.dev.dispatch(pipe, groups, &push)?;

        let raw = self.out_buf.read(self.out_size, 0);
        let count = u32::from_le_bytes(raw[0..4].try_into().unwrap());
        let hash_count = u32::from_le_bytes(raw[4..8].try_into().unwrap());
        let mut sols = Vec::new();
        for s in 0..count.min(MAX_OUTPUTS) {
            let base = 16 + (s as usize) * 9 * 4;
            let rd = |i: usize| u32::from_le_bytes(raw[base + i * 4..base + i * 4 + 4].try_into().unwrap());
            let gid = rd(0);
            let mut mix = [0u32; 8];
            for (j, m) in mix.iter_mut().enumerate() {
                *m = rd(1 + j);
            }
            sols.push(Solution {
                nonce: start_nonce + gid as u64,
                mix_words: mix,
            });
        }
        Ok((sols, hash_count as u64))
    }

    /// Find a per-dispatch nonce batch that runs ~WATCHDOG_TARGET_S. Ramps up and
    /// STOPS as soon as a dispatch reaches the target, so none trips the watchdog.
    pub fn calibrate(&mut self, header: &[u8]) -> Result<u32> {
        let mut batch = 1u64 << 13;
        let mut per_nonce: Option<f64> = None;
        while batch <= (1 << 24) {
            let t0 = Instant::now();
            let (_, hashes) = self.search(header, 0, 0, batch)?;
            let dt = t0.elapsed().as_secs_f64();
            if hashes > 0 {
                per_nonce = Some(dt / hashes as f64);
            }
            println!(
                "  calibrate batch={batch} -> {:.1} ms ({:.1} MH/s)",
                dt * 1000.0,
                hashes as f64 / dt / 1e6
            );
            if dt >= WATCHDOG_TARGET_S {
                break;
            }
            batch <<= 1;
        }
        if let Some(pn) = per_nonce {
            let safe = (WATCHDOG_TARGET_S / pn) as u64;
            let ls = self.local_size as u64;
            self.safe_batch = ((safe / ls) * ls).max(ls) as u32;
        }
        println!(
            "  safe batch = {} nonces (~{:.0} ms/dispatch)",
            self.safe_batch,
            WATCHDOG_TARGET_S * 1000.0
        );
        Ok(self.safe_batch)
    }

    /// Measure sustained hashrate (MH/s) using watchdog-safe dispatches.
    pub fn benchmark(&mut self, block: u64, seconds: f64, use_real_dag: bool) -> Result<f64> {
        if use_real_dag {
            self.ensure_epoch(block, None)?;
        } else {
            self.setup_benchmark_dag(block)?;
        }
        self.ensure_period(block)?;

        let header: Vec<u8> = (0..32u8).collect();
        println!("Calibrating safe dispatch size...");
        self.calibrate(&header)?;

        let mut start = 0u64;
        let mut total = 0u64;
        let t0 = Instant::now();
        while t0.elapsed().as_secs_f64() < seconds {
            let (_, hashes) = self.search(&header, 0, start, self.safe_batch as u64)?;
            total += hashes;
            start += self.safe_batch as u64;
        }
        let dt = t0.elapsed().as_secs_f64();
        let mhs = total as f64 / dt / 1e6;
        println!(
            "Benchmark: {:.1} Mhashes in {:.1}s = {:.2} MH/s",
            total as f64 / 1e6,
            dt,
            mhs
        );
        Ok(mhs)
    }
}
